src/protocol/assan_nrf24l01.c

```diff
@@ -0,0 +1,311 @@
+    NRF24L01_WriteReg(NRF24L01_02_EN_RXADDR, 0x01);         // Enable data pipe 0 only
+    NRF24L01_WriteReg(NRF24L01_11_RX_PW_P0, PACKET_SIZE);
+    NRF24L01_SetPower(Model.tx_power);
+
+    // Check for Beken BK2421/BK2423 chip
+    // It is done by using Beken specific activate code, 0x53
+    // and checking that status register changed appropriately
+    // There is no harm to run it on nRF24L01 because following
+    // closing activate command changes state back even if it
+    // does something on nRF24L01
+    NRF24L01_Activate(0x53); // magic for BK2421 bank switch
+    dbgprintf("Trying to switch banks\n");
+    if (NRF24L01_ReadReg(NRF24L01_07_STATUS) & 0x80) {
+        dbgprintf("BK2421 detected\n");
+        // Beken registers don't have such nice names, so we just mention
+        // them by their numbers
+        // It's all magic, eavesdropped from real transfer and not even from the
+        // data sheet - it has slightly different values
+        NRF24L01_WriteRegisterMulti(0x00, (u8 *) "\x40\x4B\x01\xE2", 4);
+        NRF24L01_WriteRegisterMulti(0x01, (u8 *) "\xC0\x4B\x00\x00", 4);
+        NRF24L01_WriteRegisterMulti(0x02, (u8 *) "\xD0\xFC\x8C\x02", 4);
+        NRF24L01_WriteRegisterMulti(0x03, (u8 *) "\x99\x00\x39\x21", 4);
+        NRF24L01_WriteRegisterMulti(0x04, (u8 *) "\xD9\x96\x82\x1B", 4);
+        NRF24L01_WriteRegisterMulti(0x05, (u8 *) "\x24\x06\x7F\xA6", 4);
+        NRF24L01_WriteRegisterMulti(0x0C, (u8 *) "\x00\x12\x73\x00", 4);
+        NRF24L01_WriteRegisterMulti(0x0D, (u8 *) "\x46\xB4\x80\x00", 4);
+        NRF24L01_WriteRegisterMulti(0x04, (u8 *) "\xDF\x96\x82\x1B", 4);
+        NRF24L01_WriteRegisterMulti(0x04, (u8 *) "\xD9\x96\x82\x1B", 4);
+    } else {
+        dbgprintf("nRF24L01 detected\n");
+    }
+    NRF24L01_Activate(0x53); // switch bank back
+}
+
+void send_packet()
+{
+    u32 temp;
+    for(u8 ch=0;ch<10;ch++)
+    {
+        temp=((s32)Channels[ch] * 0x1f1 / CHAN_MAX_VALUE + 0x5d9)<<3;
+        packet[2*ch]=temp>>8;
+        packet[2*ch+1]=temp;
+    }
+    NRF24L01_WriteReg(NRF24L01_07_STATUS, 0x70);        // Clear data ready, data sent, and retransmit
+    NRF24L01_FlushTx();
+    NRF24L01_WritePayload(packet, PACKET_SIZE);
+
+    // Check and adjust transmission power. We do this after
+    // transmission to not bother with timeout after power
+    // settings change -  we have plenty of time until next
+    // packet.
+    if (tx_power != Model.tx_power) {
+        //Keep transmit power updated
+        tx_power = Model.tx_power;
+        NRF24L01_SetPower(tx_power);
+    }
+}
+
+#define WAIT 0x80
+
+u16 ASSAN_callback()
+{
+    switch (state)
+    {
+    // Bind
+        case BIND0:
+            //Config RX @1M
+            NRF24L01_WriteReg(NRF24L01_05_RF_CH, RF_BIND_CHANNEL);
+            NRF24L01_SetBitrate(NRF24L01_BR_1M);                    // 1Mbps
+            NRF24L01_SetTxRxMode(RX_EN);
+            state++;
```

Comment:
It's a bit unclear, can we use explicit state = BIND1? It's probably justified in the case of DATA2, DATA3, but I'd avoid it as non-local - you need to check enum to figure what follows what. Especially in the transition from BIND to DATA.